_posts/2024-09-23-wai-aria.md

```diff
@@ -0,0 +1,148 @@
+
+ARIA provides a set of rules and attributes that allow developers to define more refined semantics for UI elements. When JavaScript is added to enhance user interactions, extra guidance is often needed for screen readers and other assistive technologies. ARIA enables the use of non-semantic HTML, though this doesn't mean developers should aim to write non-semantic HTML. ARIA helps accessibility tools communicate what’s happening on the page and what interactions are possible.
+
+## HTML vs. ARIA: Which Should You Use?
+
+When comparing HTML to ARIA, ARIA should always be a last resort. **If a semantic HTML element is available, use that.** For instance, there’s no need for HTML like this: `<div role="navigation">...</div>`. Instead, use the native HTML element `<nav>...</nav>`.
+
+Before using ARIA, check [the native HTML elements](https://developer.mozilla.org/en-US/docs/Web/HTML) to ensure none apply to your situation. A lot can be achieved with HTML alone.
+
+## How to Use ARIA
+
+ARIA consists of roles and attributes. Let's explore their purposes.
+
+### Roles
+
+ARIA roles make it possible to describe an element's function in a way that HTML can't. HTML elements have no role unless one is specifically assigned, except for form elements like <input type="radio"> which automatically has the role of "radio."
+
+ARIA includes roles such as `tooltip`, `tab`, and `dialog`. The roles can be divided into six categories: (1) document structure roles, (2) widget roles, (3) landmark roles, (4) live region roles, (5) window roles, and (6) abstract roles. You can find the full list of roles on [MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles).
+
+Many roles have HTML equivalents, so typically there's no need to use `role="article"` when you can use the native `<article>` element. However, in certain cases, assigning a role is necessary. For example, live region roles help alert screen readers to elements that require the user’s immediate attention.
+
+That said, it’s crucial to use `role="alert"` only when the user's immediate attention is genuinely required. Misusing it can disrupt the user experience by diverting the screen reader's focus. Just because you, as a developer, think something is urgent doesn’t mean the end user will feel the same way.
+
+### Attributes
+
+ARIA attributes modify an element’s state or properties for the accessibility tree. While ARIA attributes primarily affect the accessibility tree, they can also be used to modify the DOM, although this requires additional steps. For instance, you can apply CSS based on ARIA attributes, such as `[aria-hidden="true"] { display: none }`. However, simply adding attributes like `aria-hidden="true"` or `aria-expanded="true"` won’t impact the visual appearance of the page, but will update the accessibility tree. This allows elements to be hidden from assistive technologies or signals that one element controls another that is currently expanded.
```

Comment:
One more comment about the correctness of this.

> While ARIA attributes primarily affect the accessibility tree, they can also be used to modify the DOM, although this requires additional steps.

Just by setting an attribute (be it ARIA attribute or not) you are modifying the DOM. You are affecting the DOM in any case, but to get visual changes to the document you will need the extra CSS styles for example.